### PR TITLE
Atomic design: add select molecule to cf-forms 

### PR DESCRIPTION
**Pull request by cfarm**
Part of Atomic design issue https://github.com/cfpb/capital-framework/issues/345

Adds cf-forms from cfgov-refresh (based on @anselmbradford's update_form_elems branch) + makes select element a molecule.

This is the same changes from #350, but branching off of `prerelease-ver-4`

## Additions

- cf-forms.less from cfgov-refresh
- select element CSS from cfpb/cfgov-refresh#1273: New inputs and textareas
- markup for select elements in `usage.md`

## Testing

- npm link this and check out the following branch on capital-framework-sandbox to test: https://github.com/cfarm/capital-framework-sandbox/tree/atomic/cf-forms

## Review

- @jimmynotjim or @KimberlyMunoz 

## Screenshots
screen shot 2016-07-20 at 11 32 01 am
screen shot 2016-07-20 at 11 31 52 am

## Checklist

* [ ] Changes are limited to a single goal (no scope creep)
* [ ] Code can be automatically merged (no conflicts)
* [ ] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
* [ ] Passes all existing automated tests
* [ ] New functions include new tests
* [ ] New functions are documented (with a description, list of inputs, and expected output)
* [ ] Placeholder code is flagged
* [ ] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)

### CODE REVIEW

File: src/cf-forms/src/cf-forms.less [hunk 1]
```diff
@@ -219,6 +219,663 @@ textarea {
 
 }
 
+
+
+
+
+/* ==========================================================================
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: We don't need topdoc comments any more, everything is in the usage files now.

File: src/cf-forms/src/cf-forms.less [hunk 2]
```diff
@@ -219,6 +219,663 @@ textarea {
...
+          viewport reaches 600px they transform into columns with one half the
+          width of the container. When the viewport reaches 768px their width
+          updates to one third of the container."
+  tags:
+    - cfgov-forms
+*/
+
+.form-l {
+    &__flush {
+        .respond-to-min(@bp-lg-min, {
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: Minor, add spaces inside parens

File: src/cf-forms/src/cf-forms.less [hunk 3]
```diff
@@ -219,6 +219,663 @@ textarea {
...
+            .grid_nested-col-group();
+        });
+    }
+    &__float {
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: Minor, add a line break between rules for consistency

File: src/cf-forms/src/cf-forms.less [hunk 4]
```diff
@@ -219,6 +219,663 @@ textarea {
...
+                    .form-l-inset_container
+                        input[type="radio"].cf-input
+                        label.cf-input_label
+        notes:
+          - "Use the markup structure above to add radio button .cf-input elements."
+    tags:
+      - cf-forms
+*/
+
+@input-active:          #0072ce; // pacific.
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: Colors should be "unthemed". So far we've just been picking reasonable defaults from https://standards.usa.gov/colors/

File: src/cf-forms/src/cf-forms.less [hunk 5]
```diff
@@ -219,6 +219,663 @@ textarea {
...
+
+@input-bg-disabled:     #5a5d61; // gray;
+@input-border-disabled: #b4b5b6; // gray-40;
+
+// Insets
+@input-inset-bg:        #d2d3d5; // gray-20;
+@input-inset-selected:  #d6e8fa; // pacific-20;
+
+@checkbox-width: unit( 18px / @base-font-size-px, em);
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: Left align these two values with the others above.

File: src/cf-forms/src/cf-forms.less [hunk 6]
```diff
@@ -219,6 +219,663 @@ textarea {
...
+        .u-visually-hidden();
+
+        .lt-ie9 & {
+            height: 20px;
+            width: 20px;
+            border: 0;
+            float: left;
+            margin: 1em;
+            position: static;
+            width: auto;
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: There's two widths in here

File: src/cf-forms/src/cf-forms.less [hunk 7]
```diff
@@ -219,6 +219,663 @@ textarea {
...
+
+
+    &[type="checkbox"],
+    &[type="radio"]{
+        .u-visually-hidden();
+
+        .lt-ie9 & {
+            height: 20px;
+            width: 20px;
+            border: 0;
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: Minor: Match FEWD standards CSS ordering.

File: src/cf-forms/src/cf-forms.less [hunk 8]
```diff
@@ -219,6 +219,663 @@ textarea {
...
+              .form-l-inset_container
+        notes:
+          - ".form-l-inset_container elements have larger target areas for easier user interaction."
+    tags:
+      - cf-forms
+*/
+
+.form-l-inset {
+    border-right-width: 2.5px;
+    + .form-l_col-1-2.form-l-inset:nth-child(odd) { border-left-width: 2.5px; }
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: Add line breaks between rules for better readability and put rule contents on new lines for consistency.

File: src/cf-forms/usage.md [hunk 9]
```diff
@@ -324,10 +337,151 @@ Provides sizeable margins between groups of form elements.
...
+            <label class="cf-input_label"
+                   for="input_id_dogs">
+                Dogs
+            </label>
+        </div>
+    </div>
+</fieldset>
+
+```
+
```
Comment (jimmynotjim) [inline src/cf-forms/usage.md]: Add markup from above here

Comment (cfarm) [inline src/cf-forms/usage.md]: By "markup from above" do you mean the old Top Docs examples?

**Comment by cfarm** [comment]
Made minor and style fixes, still need to do the following:

- Colors should be "unthemed". So far we've just been picking reasonable defaults from https://standards.usa.gov/colors/
- Add topdocs markup to usage.md

File: src/cf-forms/usage.md (hunk 9 above)
Comment (jimmynotjim) [inline src/cf-forms/usage.md]: Sorry, you have markup on L:351-386 which will be rendered. Add the same markup to the code blocks so folks know what it is. The same is happening below, but there's no empty code block like you have here.

File: src/cf-forms/src/cf-forms.less (hunk 4 above)
Comment (cfarm) [inline src/cf-forms/src/cf-forms.less]: So where do the actual colors go for use in our actual projects?

Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: In the projects `cf-theme-overrides.less` file.

Comment (cfarm) [inline src/cf-forms/src/cf-forms.less]: So there's no canonical, central source that defines those colors people use? Is there a `cf-theme-overrides.less` in Capital Framework where I can add those? Otherwise people won't know which colors to add to their projects.

**Comment by cfarm** [comment]
@jimmynotjim ready for your review again. I added the theme colors to a theme-overrides stylesheet - it seems like a bad idea not to have this somewhere for other projects to grab.

File: src/cf-forms/src/cf-forms.less [hunk 10]
```diff
@@ -219,6 +219,340 @@ textarea {
...
+@input-active:          #046b99;
+
+@input-bg-disabled:     #5b616b;
+@input-border-disabled: #aeb0b5;
+
+// Insets
+@input-inset-bg:        #d6d7d9;
+@input-inset-selected:  #e1f3f8;
+
+@checkbox-width:        unit( 18px / @base-font-size-px, em);
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: Minor, missing spaces in closing parens.

File: src/cf-forms/src/cf-forms.less [hunk 11]
```diff
@@ -219,6 +219,340 @@ textarea {
...
+        cursor: pointer;
+        margin-bottom: 0.5em;
+    }
+
+    &[type="checkbox"] + label::before,
+    &[type="radio"] + label::before {
+        background: white;
+        content: '\a0';
+        display: inline-block;
+        // margin-top: unit( 2px / @base-font-size-px, em);
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: Can this be removed?

File: src/cf-forms/src/cf-forms.less [hunk 12]
```diff
@@ -219,6 +219,340 @@ textarea {
...
+        box-shadow: 0 0 0 2px @input-active;
+    }
+
+    &[type="checkbox"]:disabled + label {
+        color: @input-border-disabled;
+    }
+
+    &[type="checkbox"]:disabled + label::before,
+    &[type="radio"]:disabled + label::before {
+      background: @input-bg-disabled;
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: Minor, 2-space indentation here, 4-space elsewhere.

File: src/cf-themes-overrides.less [hunk 13]
```diff
@@ -0,0 +1,17 @@
+/* ==========================================================================
```
Comment (jimmynotjim) [inline src/cf-themes-overrides.less]: Theme overrides don't live in Capital Framework, they live in the [Yeoman generator](https://github.com/cfpb/generator-cf/blob/master/app/templates/src/static/css/cf-theme-overrides.less). Could you move these there? We'll address the whole theming thing after the Atomic work is complete.

File: src/cf-forms/src/cf-forms.less [hunk 14]
```diff
@@ -219,6 +219,340 @@ textarea {
...
+@input-border-disabled: #aeb0b5;
+
+// Insets
+@input-inset-bg:        #d6d7d9;
+@input-inset-selected:  #e1f3f8;
+
+@checkbox-width:        unit( 18px / @base-font-size-px, em);
+@radio-width:           unit( 14px / @base-font-size-px, em);
+
+.cf-input {
```
Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: I thought this PR was just for the select elem? Should it be for both select and inputs? Just want to make sure you and @KimberlyMunoz aren't duplicating efforts and having to work out conflicts later.

Comment (KimberlyMunoz) [inline src/cf-forms/src/cf-forms.less]: My stuff is namescoped differently and the checkbox specific stuff mostly copied from cfgov-refresh. I think ideally we should be doing one input at a time in CF, but I can clean this up when this is merged and I work on getting checkboxes in CF.

Comment (jimmynotjim) [inline src/cf-forms/src/cf-forms.less]: Sounds good. Sorry I didn't notice this was extra the first time around.

File: src/cf-themes-overrides.less (hunk 13 above)
Comment (cfarm) [inline src/cf-themes-overrides.less]: I put these in https://github.com/cfpb/generator-cf/pull/123

**Comment by cfarm** [comment]
Fixes made. Can I get a thumbs uppppppppp

**Comment by cfarm** [comment]
Questions: do i need to PR the capital-framework-sandbox branch for this too?

And why is Travis failing?

File: src/cf-forms/src/cf-forms.less [hunk 15]
```diff
@@ -219,6 +219,340 @@ textarea {
...
+    .form-l_col-2-3 {
+        .grid_column(8);
+    }
+
+    .form-l_col-1 {
+        .grid_column(12);
+    }
+});
+
+//
```
Comment (KimberlyMunoz) [inline src/cf-forms/src/cf-forms.less]: Lines 318-490 could technically be removed and added later. If we want to have basic checkbox styles and fix them later, we should still remove everything in the `.form-l-inset` class since that's all refactored away in #384

File: src/cf-forms/usage.md [hunk 16]
```diff
@@ -354,10 +367,277 @@ Provides sizeable margins between groups of form elements.
...
+           value="This is some really long text to make sure that the button doesn't overlap the content in such a way that this input becomes unusable."
+           title="Test input">
     <button class="btn btn__super btn__link btn__secondary">
         Clear
         <span class="cf-icon cf-icon-delete"></span>
     </button>
 </div>
 ```
+
+### Checkboxes with large target areas
```
Comment (KimberlyMunoz) [inline src/cf-forms/usage.md]: I'd remove up to line 534 since that's refactored in #384. Otherwise we can make a note to fix in another PR.

File: src/cf-forms/usage.md [hunk 17]
```diff
@@ -354,10 +367,277 @@ Provides sizeable margins between groups of form elements.
...
+
+### Disabled select
+
+<div class="form-l_col form-l_col-1">
+    <label class="form-label-header"
+           for="select_example">
+            Disabled select example
+    </label>
+    <div class="m-select">
+        <select id="select_example" disabled>
```
Comment (KimberlyMunoz) [inline src/cf-forms/usage.md]: A unique id like `select_example__disabled`

Comment (KimberlyMunoz) [inline src/cf-forms/usage.md]: Also, the same issue in L:616

File: src/cf-forms/src/cf-forms.less [hunk 18]
```diff
@@ -219,6 +219,340 @@ textarea {
...
+.no-csspointerevents .m-select {
+    &:after {
+        height: 0;
+        width: 0;
+        border: 0;
+        content: '';
+    }
+}
+
+//
```
Comment (KimberlyMunoz) [inline src/cf-forms/src/cf-forms.less]: This top doc comment isn't needed anymore 🎉

File: src/cf-forms/src/cf-forms.less [hunk 19]
```diff
@@ -219,6 +219,340 @@ textarea {
...
+        color: @gray-80;
+        background-color: @gray-10;
+    }
+
+    select option:disabled {
+        color: @gray-80;
+    }
+
+    &:after {
+        position: absolute;
```
Comment (KimberlyMunoz) [inline src/cf-forms/src/cf-forms.less]: I think I might have originally wrote this, but positioning variables go after box-model variables. Those should get rearranged.

Comment (KimberlyMunoz) [inline src/cf-forms/src/cf-forms.less]: I made this same mistake. Positioning variables go after box-model variables. Those should get rearranged.

File: src/cf-forms/src/cf-forms.less [hunk 20]
```diff
@@ -219,6 +219,340 @@ textarea {
...
+}
+
+//
+// m-select menu
+//
+
+.m-select {
+    position: relative;
+    margin-top: 0.5em;
+    border: 1px solid @gray-40;
```
Comment (KimberlyMunoz) [inline src/cf-forms/src/cf-forms.less]: I think there's a variable like `@input-border` for this and all the other color variables in this molecule. We should use those existing theming variables where ever possible.

File: src/cf-forms/src/cf-forms.less [hunk 21]
```diff
@@ -219,6 +219,340 @@ textarea {
 
 }
 
+.input-group {
```
Comment (KimberlyMunoz) [inline src/cf-forms/src/cf-forms.less]: Do we use the `.input-group` class?

**Comment by KimberlyMunoz** [comment]
>do i need to PR the capital-framework-sandbox branch for this too?

That'd be awesome. Just to make sure that none of the other form work affects this stuff from working properly.

Also, Travis is having an issue with our unit tests. It's fine since your work didn't touch them but we need to look into it later.

File: src/cf-forms/src/cf-forms.less (hunk 21 above)
Comment (KimberlyMunoz) [inline src/cf-forms/src/cf-forms.less]: BTWs, this is more of a just asking to see if it makes sense to remove. It's fine to leave in.

File: src/cf-themes-overrides.less [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+// Checkboxes and radio buttons
+//
+
+@input-active:          #0072ce; // pacific.
+
+@input-bg-disabled:     #5a5d61; // gray;
+@input-border-disabled: #b4b5b6; // gray-40;
+
+// Insets
+@input-inset-bg:        #d2d3d5; // gray-20;
```
Comment (cfarm) [inline src/cf-themes-overrides.less]: if I delete inset code I can remove these for now.

File: src/cf-themes-overrides.less [hunk 23]
```diff
@@ -0,0 +1,17 @@
+/* ==========================================================================
+   Capital Framework
+   Form Element Styling
+   ========================================================================== */
+
+//
+// Checkboxes and radio buttons
+//
+
+@input-active:          #0072ce; // pacific.
```
Comment (cfarm) [inline src/cf-themes-overrides.less]: use existing variables in cf-forms

File: src/cf-forms/src/cf-forms.less (hunk 20 above)
Comment (cfarm) [inline src/cf-forms/src/cf-forms.less]: are you talking about variables in cf-generator? they are not in sync with the updated form designs: https://github.com/cfpb/generator-cf/blob/f86d1d849bc355b0e40f332623457194c5a266f6/app/templates/src/static/css/cf-theme-overrides.less#L47
so i need to make that change there, too, right?

Comment (cfarm) [inline src/cf-forms/src/cf-forms.less]: just a note that i'm adding a new variable for the select-border since its unique, will update cf-generator to reflect

Comment (KimberlyMunoz) [inline src/cf-forms/src/cf-forms.less]: Hmmm... Maybe just leave it since it seems like a bigger haul than I initially thought and it's something that can be patched up soon.

**Comment by cfarm** [comment]
New color variables added in https://github.com/cfpb/generator-cf/pull/124

File: src/cf-forms/src/cf-forms.less (hunk 21 above)
Comment (cfarm) [inline src/cf-forms/src/cf-forms.less]: this was equivalent to the new o-form-input-group: https://github.com/cfpb/capital-framework/pull/384/files#diff-e37310c249b73d03f5dde096d70e3aa9

i'll remove

**Comment by cfarm** [comment]
This should be ready to merge now.

**Comment by KimberlyMunoz** [comment]
Lovely! :+1:
